Bug fix: `docker run -i --restart always` hangs

WeiZhang555 opened the pull request:
Bug fix: docker run -i --restart always hangs on exit.

e.g.

```
$ docker run -i --restart always busybox sh
pwd
/
exit 11

<...hang...>
```

This is because Attach(daemon side) and Run(client side) both hangs on
`WaitStop`, if container is restarted too quickly, wait won't have chance
to get stop signal.

Besides that, restarting state should be considered stopped while
calling Wait, `docker wait` a restarting container should return
immediately too.

Signed-off-by: Zhang Wei zhangwei555@huawei.com

WeiZhang555 committed:
Bug fix: `docker run -i --restart always` hangs.

e.g.
```
$ docker run -i --restart always busybox sh
pwd
/
exit 11

<...hang...>
```

This is because Attach(daemon side) and Run(client side) both hangs on
WaitStop, if container is restarted too quickly, wait won't have chance
to get exit signal.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
```
02:41:51 Step 54 : COPY . /go/src/github.com/docker/docker
02:41:52 devmapper: Error mounting '/dev/mapper/docker-202:1-75497728-1ccd825754a65bb96750fec682aedaa93e365a05845085bdd2618d05ea9072a1-init' on '/var/lib/docker/devicemapper/mnt/1ccd825754a65bb96750fec682aedaa93e365a05845085bdd2618d05ea9072a1-init': no space left on device
02:41:52 Unable to find image 'docker:bc21821-exp' locally
```

Is test machine's disk full??

runcom commented:
is this a regression? caused by?

Fix isn't working for me (I can reproduce though)

WeiZhang555 commented:
@runcom   Yep, this is a regression. 

I think it's caused by some race problem. Attach will Wait container to stop after stream is closed (https://github.com/docker/docker/blob/master/daemon/attach.go#L126 and https://github.com/docker/docker/blob/master/api/client/run.go#L285 ), but if container is started with some restart policy, and if it's restarted too quickly, the `wait` operation may hang.

If the root cause I mentioned is correct, I think I can understand why the fix isn't working for you, maybe your container is already restarted successfully when `WaitStop` is invoked and container already enter state running, so the `wait` still hangs.

It seems this commit doesn't fix the bug completely, maybe we should add a timeout? I'll do more investigations.

WeiZhang555 commented:
@runcom I made a little bit modification, and believe it can fix this bug completely. Correct me if it doesn't.

Test machines have some unrelated problems.

LK4D4 commented:
@WeiZhang555 docker-py failure looks valid

WeiZhang555 commented:
Oh, sorry, I thought it's unrelated. I'll check.

WeiZhang555 commented:
I think it's OK now :sunglasses:

WeiZhang555 commented:
@thaJeztah Could you help restart `windowsTP5` test? I believe it will be green after retart :sunglasses:

thaJeztah commented:
Hm I restarted twice; weird. Either something's wrong with the windows CI, or something genuine. Let's try once more

WeiZhang555 commented:
Thank you @thaJeztah !  It's weird, Windows CI hangs...

LK4D4 commented:
Let's try again :)

LK4D4 commented:
I think it's stuck again :/ ping @jhowardmsft @jstarks

jhowardmsft commented:
Looks like this is not fixed on Windows.

```
time="2016-05-23T20:36:24.437978500Z" level=info msg="=== BEGIN goroutine stack dump ===
goroutine 21 [running]:
github.com/docker/docker/pkg/signal.DumpStacks()
	c:/go/src/github.com/docker/docker/pkg/signal/trap.go:67 +0x9f
github.com/docker/docker/daemon.setupDumpStackTrap.func1()
	c:/go/src/github.com/docker/docker/daemon/debugtrap_windows.go:26 +0x2ae
created by github.com/docker/docker/daemon.setupDumpStackTrap
	c:/go/src/github.com/docker/docker/daemon/debugtrap_windows.go:29 +0x32

goroutine 1 [chan receive, 137 minutes]:
main.(*DaemonCli).start(0xc082338300, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:293 +0x1f8e
main.main()
	c:/go/src/github.com/docker/docker/cmd/dockerd/docker.go:68 +0x3a1

goroutine 17 [syscall, 137 minutes, locked to thread]:
runtime.goexit()
	c:/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 34 [select, 97 minutes]:
github.com/docker/libnetwork.(*controller).watchLoop(0xc082387080)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/store.go:431 +0x141
created by github.com/docker/libnetwork.(*controller).startWatch
	C:/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/store.go:448 +0x119

goroutine 18 [syscall, 137 minutes]:
os/signal.loop()
	c:/go/src/os/signal/signal_unix.go:22 +0x1f
created by os/signal.init.1
	c:/go/src/os/signal/signal_unix.go:28 +0x3e

goroutine 20 [select, 97 minutes]:
github.com/Microsoft/go-winio.(*win32PipeListener).listenerRoutine(0xc082336460)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/pipe.go:267 +0x44c
created by github.com/Microsoft/go-winio.ListenPipe
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/pipe.go:350 +0x4a0

goroutine 35 [chan receive, 3 minutes]:
github.com/docker/docker/daemon.(*Daemon).execCommandGC(0xc082380000)
	c:/go/src/github.com/docker/docker/daemon/exec.go:216 +0x95
created by github.com/docker/docker/daemon.NewDaemon
	c:/go/src/github.com/docker/docker/daemon/daemon.go:830 +0x349d

goroutine 22 [syscall, 137 minutes, locked to thread]:
syscall.Syscall(0x7fff7f5c43a0, 0x2, 0x334, 0xffffffff, 0x0, 0x24, 0xc0822a5f18, 0x1)
	c:/go/src/runtime/syscall_windows.go:128 +0x5c
syscall.WaitForSingleObject(0x334, 0xffffffff, 0xc0822a5f18, 0x0, 0x0)
	c:/go/src/syscall/zsyscall_windows.go:690 +0x6f
main.(*DaemonCli).setupConfigReloadTrap.func1(0xc082338300)
	c:/go/src/github.com/docker/docker/cmd/dockerd/daemon_windows.go:58 +0x2a9
created by main.(*DaemonCli).setupConfigReloadTrap
	c:/go/src/github.com/docker/docker/cmd/dockerd/daemon_windows.go:62 +0x3c

goroutine 23 [chan receive, 137 minutes]:
github.com/docker/docker/api/server.(*Server).serveAPI(0xc082215320, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/api/server/server.go:93 +0x1a5
github.com/docker/docker/api/server.(*Server).Wait(0xc082215320, 0xc082475020)
	c:/go/src/github.com/docker/docker/api/server/server.go:182 +0x32
created by main.(*DaemonCli).start
	c:/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:281 +0x1ee7

goroutine 24 [chan receive, 137 minutes]:
github.com/docker/docker/pkg/signal.Trap.func1(0xc082475080, 0xc0824b9600)
	c:/go/src/github.com/docker/docker/pkg/signal/trap.go:29 +0x78
created by github.com/docker/docker/pkg/signal.Trap
	c:/go/src/github.com/docker/docker/pkg/signal/trap.go:55 +0x221

goroutine 25 [chan receive, 97 minutes]:
github.com/Microsoft/go-winio.(*win32PipeListener).Accept(0xc082336460, 0x0, 0x0, 0x0, 0x0)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/pipe.go:371 +0x132
net/http.(*Server).Serve(0xc082341440, 0x3839b68, 0xc082336460, 0x0, 0x0)
	c:/go/src/net/http/server.go:1887 +0xba
github.com/docker/docker/api/server.(*HTTPServer).Serve(0xc082344320, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/api/server/server.go:112 +0x4e
github.com/docker/docker/api/server.(*Server).serveAPI.func1(0xc082475140, 0xc082344320)
	c:/go/src/github.com/docker/docker/api/server/server.go:85 +0x144
created by github.com/docker/docker/api/server.(*Server).serveAPI
	c:/go/src/github.com/docker/docker/api/server/server.go:89 +0x13e

goroutine 26 [syscall, 97 minutes, locked to thread]:
syscall.Syscall6(0x7fff7f5b6f90, 0x5, 0x354, 0xc08237df54, 0xc08237df58, 0xc08237df60, 0xffffffff, 0x0, 0xc08237df78, 0x1, ...)
	c:/go/src/runtime/syscall_windows.go:139 +0x5c
github.com/Microsoft/go-winio.getQueuedCompletionStatus(0x354, 0xc08237df54, 0xc08237df58, 0xc08237df60, 0xc0ffffffff, 0x0, 0x0)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/zsyscall.go:76 +0x99
github.com/Microsoft/go-winio.ioCompletionProcessor(0x354)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/file.go:127 +0x7c
created by github.com/Microsoft/go-winio.initIo
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/file.go:55 +0xb3

goroutine 19235 [semacquire, 97 minutes]:
sync.runtime_Syncsemacquire(0xc08256c640)
	c:/go/src/runtime/sema.go:237 +0x20f
sync.(*Cond).Wait(0xc08256c630)
	c:/go/src/sync/cond.go:62 +0xa2
io.(*pipe).read(0xc08256c600, 0xc082ade000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	c:/go/src/io/pipe.go:52 +0x2d9
io.(*PipeReader).Read(0xc0828d2910, 0xc082ade000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)
	c:/go/src/io/pipe.go:134 +0x57
io.copyBuffer(0x20569c0, 0xc08272d240, 0x2055698, 0xc0828d2910, 0xc082ade000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	c:/go/src/io/io.go:381 +0x24e
io.Copy(0x20569c0, 0xc08272d240, 0x2055698, 0xc0828d2910, 0x419986, 0x0, 0x0)
	c:/go/src/io/io.go:351 +0x6b
github.com/docker/docker/daemon.(*Daemon).AttachStreams.func2.1(0x20569c0, 0xc08272d240, 0x2055698, 0xc0828d2910, 0xc082f0bb80, 0x40, 0xc082f0bbc0)
	c:/go/src/github.com/docker/docker/daemon/monitor.go:132 +0x80
created by github.com/docker/docker/daemon.(*Daemon).AttachStreams.func2
	c:/go/src/github.com/docker/docker/daemon/monitor.go:136 +0xb6

goroutine 19236 [semacquire, 97 minutes]:
sync.runtime_Syncsemacquire(0xc08256c700)
	c:/go/src/runtime/sema.go:237 +0x20f
sync.(*Cond).Wait(0xc08256c6f0)
	c:/go/src/sync/cond.go:62 +0xa2
io.(*pipe).read(0xc08256c6c0, 0xc082ae6000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	c:/go/src/io/pipe.go:52 +0x2d9
io.(*PipeReader).Read(0xc0828d2920, 0xc082ae6000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)
	c:/go/src/io/pipe.go:134 +0x57
io.copyBuffer(0x20569c0, 0xc08272d220, 0x2055698, 0xc0828d2920, 0xc082ae6000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	c:/go/src/io/io.go:381 +0x24e
io.Copy(0x20569c0, 0xc08272d220, 0x2055698, 0xc0828d2920, 0xc0826cb680, 0x0, 0x0)
	c:/go/src/io/io.go:351 +0x6b
github.com/docker/docker/daemon.(*Daemon).AttachStreams.func2.1(0x20569c0, 0xc08272d220, 0x2055698, 0xc0828d2920, 0xc082f0bb80, 0x40, 0xc082f0bbc0)
	c:/go/src/github.com/docker/docker/daemon/monitor.go:132 +0x80
created by github.com/docker/docker/daemon.(*Daemon).AttachStreams.func2
	c:/go/src/github.com/docker/docker/daemon/monitor.go:136 +0xb6

goroutine 19234 [semacquire, 97 minutes]:
sync.runtime_Syncsemacquire(0xc0824959c0)
	c:/go/src/runtime/sema.go:237 +0x20f
sync.(*Cond).Wait(0xc0824959b0)
	c:/go/src/sync/cond.go:62 +0xa2
io.(*pipe).read(0xc082495980, 0xc082ad6000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	c:/go/src/io/pipe.go:52 +0x2d9
io.(*PipeReader).Read(0xc0820d6c00, 0xc082ad6000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)
	c:/go/src/io/pipe.go:134 +0x57
io.copyBuffer(0x3860f48, 0xc082c3eaf0, 0x2055698, 0xc0820d6c00, 0xc082ad6000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	c:/go/src/io/io.go:381 +0x24e
io.Copy(0x3860f48, 0xc082c3eaf0, 0x2055698, 0xc0820d6c00, 0xc0820d6c00, 0x0, 0x0)
	c:/go/src/io/io.go:351 +0x6b
github.com/docker/docker/daemon.(*Daemon).AttachStreams.func1(0x2056928, 0xc082c3eaf0, 0x2055698, 0xc0828d2910, 0x2055698, 0xc0828d2920, 0xc082f0bb00, 0x3844268, 0xc0820d6c00)
	c:/go/src/github.com/docker/docker/daemon/monitor.go:116 +0xb4
created by github.com/docker/docker/daemon.(*Daemon).AttachStreams
	c:/go/src/github.com/docker/docker/daemon/monitor.go:118 +0x306

goroutine 19222 [chan receive, 97 minutes]:
github.com/Microsoft/go-winio.(*win32File).asyncIo(0xc0827f5590, 0xc082c9ffb0, 0x0, 0xc000000000, 0x0, 0x0, 0x2050168, 0xc082448b40, 0x0, 0x0, ...)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/file.go:165 +0x176
github.com/Microsoft/go-winio.connectPipe(0xc0827f5590, 0x0, 0x0)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/pipe.go:360 +0x10f
github.com/Microsoft/go-winio.(*win32PipeListener).listenerRoutine.func1(0xc082daa8a0, 0xc082362e20)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/pipe.go:265 +0x2b
created by github.com/Microsoft/go-winio.(*win32PipeListener).listenerRoutine
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/pipe.go:266 +0x1fa

goroutine 19169 [semacquire, 97 minutes]:
sync.runtime_Syncsemacquire(0xc0822826d0)
	c:/go/src/runtime/sema.go:237 +0x20f
sync.(*Cond).Wait(0xc0822826c0)
	c:/go/src/sync/cond.go:62 +0xa2
github.com/docker/docker/pkg/ioutils.(*BytesPipe).Read(0xc082282440, 0xc082a7e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/pkg/ioutils/bytespipe.go:134 +0xfe
bufio.(*Reader).fill(0xc082379ec0)
	c:/go/src/bufio/bufio.go:97 +0x1f0
bufio.(*Reader).ReadSlice(0xc082379ec0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
	c:/go/src/bufio/bufio.go:328 +0x221
bufio.(*Reader).ReadBytes(0xc082379ec0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
	c:/go/src/bufio/bufio.go:406 +0xb0
github.com/docker/docker/daemon/logger.(*Copier).copySrc(0xc082282740, 0x12e8058, 0x6, 0x3879f00, 0xc082282440)
	c:/go/src/github.com/docker/docker/daemon/logger/copier.go:53 +0x151
created by github.com/docker/docker/daemon/logger.(*Copier).Run
	c:/go/src/github.com/docker/docker/daemon/logger/copier.go:40 +0x12e

goroutine 19166 [chan receive, 97 minutes]:
github.com/Microsoft/go-winio.(*win32File).asyncIo(0xc082c3ee60, 0xc0828c2db0, 0x0, 0x0, 0x0, 0xc000000000, 0x2050168, 0xc0824b7170, 0xeb30c0, 0x0, ...)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/file.go:165 +0x176
github.com/Microsoft/go-winio.(*win32File).Read(0xc082c3ee60, 0xc082ace000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/file.go:188 +0x15f
io.copyBuffer(0x2055848, 0xc0828d2928, 0x44e0728, 0xc082c3ee60, 0xc082ace000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	c:/go/src/io/io.go:381 +0x24e
io.Copy(0x2055848, 0xc0828d2928, 0x44e0728, 0xc082c3ee60, 0xc082c3ee60, 0x0, 0x0)
	c:/go/src/io/io.go:351 +0x6b
github.com/docker/docker/libcontainerd.openReaderFromPipe.func1(0xc0828d2928, 0x2056958, 0xc082c3ee60, 0xc0828d2920)
	c:/go/src/github.com/docker/docker/libcontainerd/process_windows.go:22 +0xa0
created by github.com/docker/docker/libcontainerd.openReaderFromPipe
	c:/go/src/github.com/docker/docker/libcontainerd/process_windows.go:27 +0x1a6

goroutine 19165 [chan receive, 97 minutes]:
github.com/Microsoft/go-winio.(*win32File).asyncIo(0xc082c3eb40, 0xc0828c2d50, 0x0, 0x0, 0x0, 0xc000000000, 0x2050168, 0xc0824b7160, 0xeb30c0, 0x0, ...)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/file.go:165 +0x176
github.com/Microsoft/go-winio.(*win32File).Read(0xc082c3eb40, 0xc082cb0000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/go-winio/file.go:188 +0x15f
io.copyBuffer(0x2055848, 0xc0828d2918, 0x44e0728, 0xc082c3eb40, 0xc082cb0000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	c:/go/src/io/io.go:381 +0x24e
io.Copy(0x2055848, 0xc0828d2918, 0x44e0728, 0xc082c3eb40, 0xc082c3eb40, 0x0, 0x0)
	c:/go/src/io/io.go:351 +0x6b
github.com/docker/docker/libcontainerd.openReaderFromPipe.func1(0xc0828d2918, 0x2056958, 0xc082c3eb40, 0xc0828d2910)
	c:/go/src/github.com/docker/docker/libcontainerd/process_windows.go:22 +0xa0
created by github.com/docker/docker/libcontainerd.openReaderFromPipe
	c:/go/src/github.com/docker/docker/libcontainerd/process_windows.go:27 +0x1a6

goroutine 19167 [syscall, 97 minutes, locked to thread]:
syscall.Syscall6(0x7fff73bdc8b0, 0x4, 0xc082d60ab0, 0x3db0, 0xffffffff, 0xc0826b1b74, 0x0, 0x0, 0x41, 0x41, ...)
	c:/go/src/runtime/syscall_windows.go:139 +0x5c
github.com/Microsoft/hcsshim._waitForProcessInComputeSystem(0xc082d60ab0, 0xffffffff00003db0, 0xc0826b1b74, 0x0, 0x0)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/hcsshim/zhcsshim.go:710 +0xc6
github.com/Microsoft/hcsshim.waitForProcessInComputeSystem(0xc082f0bb80, 0x40, 0xffffffff00003db0, 0xc0826b1b74, 0x0, 0x0)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/hcsshim/zhcsshim.go:703 +0x8b
github.com/Microsoft/hcsshim.WaitForProcessInComputeSystem(0xc082f0bb80, 0x40, 0xffffffff00003db0, 0xeb2fc0, 0x0, 0x0)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/Microsoft/hcsshim/waitprocess.go:13 +0x255
github.com/docker/docker/libcontainerd.(*container).waitExit(0xc082ec8380, 0x3db0, 0x12e4bf8, 0x4, 0x1, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/libcontainerd/container_windows.go:147 +0x204
created by github.com/docker/docker/libcontainerd.(*container).start
	c:/go/src/github.com/docker/docker/libcontainerd/container_windows.go:121 +0x1029

goroutine 19221 [chan receive, 97 minutes]:
github.com/docker/docker/container.wait(0xc082782d20, 0xffffffffc4653600, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/container/state.go:109 +0x6f
github.com/docker/docker/container.(*State).WaitStop(0xc082cc3650, 0xffffffffc4653600, 0x9, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/container/state.go:132 +0xbf
github.com/docker/docker/daemon.(*Daemon).Kill(0xc082380000, 0xc082cdfd40, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/daemon/kill.go:140 +0x2c4
github.com/docker/docker/daemon.(*Daemon).cleanupContainer(0xc082380000, 0xc082cdfd40, 0x1, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/daemon/delete.go:86 +0x2cb
github.com/docker/docker/daemon.(*Daemon).ContainerRm(0xc082380000, 0xc0826def99, 0xc, 0xc0824b7070, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/daemon/delete.go:42 +0x22e
github.com/docker/docker/api/server/router/container.(*containerRouter).deleteContainers(0xc0823edc80, 0x383a8c0, 0xc0828c2ae0, 0x20551d8, 0xc0820e9600, 0xc082b588c0, 0xc0828c2870, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/api/server/router/container/container_routes.go:374 +0x1e5
github.com/docker/docker/api/server/router/container.(*containerRouter).(github.com/docker/docker/api/server/router/container.deleteContainers)-fm(0x383a8c0, 0xc0828c2ae0, 0x20551d8, 0xc0820e9600, 0xc082b588c0, 0xc0828c2870, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/api/server/router/container/container.go:66 +0x7b
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1(0x383a8c0, 0xc0828c2ae0, 0x20551d8, 0xc0820e9600, 0xc082b588c0, 0xc0828c2870, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/api/server/middleware/version.go:56 +0x9b1
github.com/docker/docker/api/server/middleware.UserAgentMiddleware.WrapHandler.func1(0x383a8c0, 0xc0828c29c0, 0x20551d8, 0xc0820e9600, 0xc082b588c0, 0xc0828c2870, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/api/server/middleware/user_agent.go:45 +0x6a4
github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1(0x3834b48, 0xc08200d250, 0x20551d8, 0xc0820e9600, 0xc082b588c0, 0xc0828c2870, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/api/server/middleware/debug.go:21 +0xd5e
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1(0x20551d8, 0xc0820e9600, 0xc082b588c0)
	c:/go/src/github.com/docker/docker/api/server/server.go:137 +0x13f
net/http.HandlerFunc.ServeHTTP(0xc0825205a0, 0x20551d8, 0xc0820e9600, 0xc082b588c0)
	c:/go/src/net/http/server.go:1422 +0x41
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0823a9900, 0x20551d8, 0xc0820e9600, 0xc082b588c0)
	C:/go/src/github.com/docker/docker/vendor/src/github.com/gorilla/mux/mux.go:98 +0x2a5
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP(0xc0824f9f50, 0x20551d8, 0xc0820e9600, 0xc082b588c0)
	c:/go/src/github.com/docker/docker/api/server/router_swapper.go:29 +0x79
net/http.serverHandler.ServeHTTP(0xc082341440, 0x20551d8, 0xc0820e9600, 0xc082b588c0)
	c:/go/src/net/http/server.go:1862 +0x1a5
net/http.(*conn).serve(0xc08289f6b0)
	c:/go/src/net/http/server.go:1361 +0xbf5
created by net/http.(*Server).Serve
	c:/go/src/net/http/server.go:1910 +0x3fd

goroutine 19168 [semacquire, 97 minutes]:
sync.runtime_Syncsemacquire(0xc0822dd1d0)
	c:/go/src/runtime/sema.go:237 +0x20f
sync.(*Cond).Wait(0xc0822dd1c0)
	c:/go/src/sync/cond.go:62 +0xa2
github.com/docker/docker/pkg/ioutils.(*BytesPipe).Read(0xc0822dd100, 0xc082a7c000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	c:/go/src/github.com/docker/docker/pkg/ioutils/bytespipe.go:134 +0xfe
bufio.(*Reader).fill(0xc082ab1ec0)
	c:/go/src/bufio/bufio.go:97 +0x1f0
bufio.(*Reader).ReadSlice(0xc082ab1ec0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
	c:/go/src/bufio/bufio.go:328 +0x221
bufio.(*Reader).ReadBytes(0xc082ab1ec0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
	c:/go/src/bufio/bufio.go:406 +0xb0
github.com/docker/docker/daemon/logger.(*Copier).copySrc(0xc082282740, 0x12e8068, 0x6, 0x3879f00, 0xc0822dd100)
	c:/go/src/github.com/docker/docker/daemon/logger/copier.go:53 +0x151
created by github.com/docker/docker/daemon/logger.(*Copier).Run
	c:/go/src/github.com/docker/docker/daemon/logger/copier.go:40 +0x12e

=== END goroutine stack dump ===" 
```

WeiZhang555 commented:
Weird. I have no idea why it doesn't work on Windows, I'll take a look into this.

WeiZhang555 commented:
I don't understand, I guess this is a test case problem but not the fix itself, there's a timeout in the test case so I don't expect it to hang, so where does it hang? 

Windows TP5
```
19:13:07 Build timed out (after 200 minutes). Marking the build as failed.
19:13:07 Set build name.
19:13:07 New build name is '#22777'
19:13:07 Variable with name 'BUILD_DISPLAY_NAME' already exists, current value: '#22777', new value: '#22777'
19:13:07 Build was aborted
19:13:07 [PostBuildScript] - Execution post build scripts.
19:13:08 [docker] $ sh -xe C:\Users\jenkins\hudson9115625290377212869.sh
19:13:08 + set +e
19:13:08 + set +x
19:13:08 INFO: Non-base image count on control daemon to delete is 1
19:13:08 Deleted: sha256:cccb85cc68c5b9221754cd3c4e11f0c3e29a70d5cf9a3caf2d46a94c540fbe22
19:13:08 Deleted: sha256:42396e540801496fd2def19d70489c8b9fdd23b5fecf9e2fa7a5b0dee119e04d
19:13:08 INFO: Killing daemon with PID 4240
19:13:08 SUCCESS: The process with PID 4240 (child process of PID 11876) has been terminated.
19:13:08 PASS: docker_cli_run_test.go:1813: DockerSuite.TestRunInteractiveWithRestartPolicy	3.207s
```

win2lin fails to receive output of container command:
```
16:27:38 ----------------------------------------------------------------------
16:27:38 FAIL: docker_cli_run_test.go:1813: DockerSuite.TestRunInteractiveWithRestartPolicy
16:27:38 
16:27:38 docker_cli_run_test.go:1831:
16:27:38     c.Assert(err, checker.IsNil)
16:27:38 ... value *errors.errorString = &errors.errorString{s:"EOF"} ("EOF")
16:27:38 
16:27:39 
16:27:39 ----------------------------------------------------------------------
```

@jhowardmsft can you give me some hints?

jhowardmsft commented:
I *think* what's happening here is that there's a race between the test, test cleanup, and the restart logic. Probably because Windows is slower to start/stop containers currently than Linux. The log seems to indicate that the container started, did the echo, and the exit caused it to shutdown. All successful. But the restart means the container will restart. The test ends, and goes into the regular cleanup code at the end of each test. One thing cleanup does it does is an API equivalent to docker rm -f on containers. That's where the race is - the container is restarting during the rm -f, and rm -f doesn't have adequate guards in place and thinks on entry that the container isn't running, but it changes to running half way through. I'm pretty sure the same race exists on Linux, just the timing window is much smaller due to containers starting/stopping faster.

WeiZhang555 commented:
@jhowardmsft It works! Thank you very much for help me out !

jhowardmsft commented:
Yay :) I've restarted it just for my own peace of mind that the fix is good.

WeiZhang555 commented:
All green now :palm_tree:  PTAL

LK4D4 commented:
LGTM
ping @jhowardmsft

jhowardmsft commented:
Passed twice so fix looks good from a Window side.

LK4D4 commented:
ping @docker/core-engine-maintainers

WeiZhang555 commented:
Rebased and updated.
I just introduced the usage of `detach` event and make a little more extra changes here,  so please take another look at this.  
@LK4D4 @runcom   /cc: @cpuguy83 @vdemeester @jhowardmsft

WeiZhang555 commented:
ping @runcom @vdemeester @cpuguy83 could you take a look at this? I think we'd better fix this before 1.12 because it's a regression.

*Note*: Janky actually passed but github notification is missing.

vdemeester commented:
LGTM 🐸

runcom commented:
LGTM
